pkg/gatewayserver/io/basicstationlns/basicstationlns_test.go

```diff
@@ -0,0 +1,1030 @@
+			GatewayID:      registeredGatewayID.GatewayID,
+			AuthToken:      registeredGatewayToken,
+			AuthHeader:     "Bearer",
+			ErrorAssertion: nil,
+		},
+		{
+			Name:       "RegisteredGatewayAndInValidKey",
+			GatewayID:  registeredGatewayID.GatewayID,
+			AuthToken:  "invalidToken",
+			AuthHeader: "Authorization",
+			ErrorAssertion: func(err error) bool {
+				return err == websocket.ErrBadHandshake
+			},
+		},
+		{
+			Name:           "RegisteredGatewayAndNoKey",
+			GatewayID:      registeredGatewayID.GatewayID,
+			ErrorAssertion: nil,
+		},
+		{
+			Name:       "UnregisteredGateway",
+			GatewayID:  "eui-1122334455667788",
+			AuthToken:  registeredGatewayToken,
+			AuthHeader: "Authorization",
+			ErrorAssertion: func(err error) bool {
+				return err == websocket.ErrBadHandshake
+			},
+		},
+	} {
+		t.Run(fmt.Sprintf("%s", tc.Name), func(t *testing.T) {
+			a := assertions.New(t)
+			h := http.Header{}
+			h.Set(tc.AuthHeader, tc.AuthToken)
+			conn, _, err := websocket.DefaultDialer.Dial("ws://localhost:8100/api/v3/gs/io/basicstation/traffic/"+tc.GatewayID, h)
+			if err != nil {
+				if tc.ErrorAssertion == nil || !a.So(tc.ErrorAssertion(err), should.BeTrue) {
+					t.Fatalf("Unexpected error: %v", err)
+				}
+			} else if tc.ErrorAssertion != nil {
+				t.Fatalf("Expected error")
+			}
+			if conn != nil {
+				conn.Close()
+			}
+		})
+	}
+
+}
+
+func TestClientSideTLS(t *testing.T) {
+	// TODO: https://github.com/TheThingsNetwork/lorawan-stack/issues/558
+}
+func TestDiscover(t *testing.T) {
+	ctx := log.NewContext(test.Context(), test.GetLogger(t))
+
+	c := component.MustNew(test.GetLogger(t), &component.Config{
+		ServiceBase: config.ServiceBase{
+			HTTP: config.HTTP{
+				Listen: ":8100",
+			},
+		},
+	})
+	gs := mock.NewServer()
+	srv := New(ctx, gs)
+	c.RegisterWeb(srv)
+	test.Must(nil, c.Start())
+	defer c.Close()
+
+	// Invalid Endpoints
+	for i, tc := range []struct {
+		URL string
+	}{
+		{
+			URL: "ws://localhost:8100/router-info",
+		},
+		{
+			URL: discoveryEndPoint + "/router-58a0:cbff:fe80:f8",
+		},
+		{
+			URL: discoveryEndPoint + "/eui-0101010101010101",
+		},
+	} {
+		t.Run(fmt.Sprintf("InvalidDiscoveryEndPoint/%d", i), func(t *testing.T) {
+			a := assertions.New(t)
+			_, res, err := websocket.DefaultDialer.Dial(tc.URL, nil)
+			if res.StatusCode != http.StatusNotFound {
+				t.Fatalf("Unexpected response received: %v", res.Status)
+			}
+			if !a.So(err, should.Equal, websocket.ErrBadHandshake) {
+				t.Fatalf("Connection failed: %v", err)
+			}
+		})
+	}
+
+	// Test Queries
+	for i, tc := range []struct {
+		Query    interface{}
+		Response messages.DiscoverResponse
+	}{
+		{
+			Query:    messages.DiscoverQuery{},
+			Response: messages.DiscoverResponse{Error: "Empty router EUI provided"},
+		},
+		{
+			Query:    struct{}{},
+			Response: messages.DiscoverResponse{Error: "Empty router EUI provided"},
+		},
+		{
+			Query: struct {
+				EUI string `json:"route"`
```

Comment:
Is the test here that `route` is not `router`?

Reply:
Yes 👍 . Added comments